# Data Cleaning: Company Employee Dataset


In [1]:
import pandas as pd

df = pd.read_csv('Day11_Messy_Company_Employee_Dataset.csv')
df.head()

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,NaN,Data Scientist,48.0,Other,108371.0,13.2,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.2,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,NaN,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.9,2018-07-24,Hyderabad,4.0,Hybrid


In [2]:
print('Shape:', df.shape)
print('\nData Types:')
print(df.dtypes)

print('\nMissing Values:')
print(df.isnull().sum())

print('\nTotal Missing Values:', df.isnull().sum().sum())
print('\nDuplicate Records:', df.duplicated().sum())

Shape: (157, 12)

Data Types:
Employee_ID           object
Employee_Name         object
Department            object
Job_Title             object
Age                  float64
Gender                object
Annual_Salary        float64
Experience_Years     float64
Joining_Date          object
City                  object
Performance_Score    float64
Work_Mode             object
dtype: object

Missing Values:
Employee_ID          0
Employee_Name        0
Department           5
Job_Title            0
Age                  4
Gender               5
Annual_Salary        5
Experience_Years     3
Joining_Date         0
City                 5
Performance_Score    3
Work_Mode            2
dtype: int64

Total Missing Values: 32

Duplicate Records: 7


In [3]:
for col in ['Department', 'Gender', 'City', 'Work_Mode']:
    print('\n', col)
    print(df[col].value_counts(dropna=False))


 Department
Department
Marketing           23
Engineering         21
Customer Success    20
Finance             19
Sales               19
Operations          17
Human Resources     16
Data & Analytics    15
NaN                  5
ENGINEERING          1
engineering          1
Name: count, dtype: int64

 Gender
Gender
Male      75
Female    68
Other      7
NaN        5
female     1
MALE       1
Name: count, dtype: int64

 City
City
Chandigarh    24
Pune          23
Jaipur        18
Mumbai        15
Hyderabad     14
Srinagar      13
Bengaluru     12
Delhi         11
Chennai       10
Kolkata        9
NaN            5
delhi          1
DELHI          1
Delhi          1
Name: count, dtype: int64

 Work_Mode
Work_Mode
Remote    56
Hybrid    50
Office    47
NaN        2
REMOTE     1
remote     1
Name: count, dtype: int64


In [4]:
## Remove Duplicate Records
df = df.drop_duplicates().copy()
print('Rows after removing duplicates:', len(df))
print('Remaining duplicates:', df.duplicated().sum())

Rows after removing duplicates: 150
Remaining duplicates: 0


In [5]:
## Standardize Inconsistent Entries
for col in ['Department', 'Gender', 'City', 'Work_Mode']:
    df[col] = df[col].astype('string').str.strip()

df['Department'] = df['Department'].replace({
    'ENGINEERING': 'Engineering',
    'engineering': 'Engineering'
})

df['Gender'] = df['Gender'].replace({
    'female': 'Female',
    'MALE': 'Male'
})

df['City'] = df['City'].replace({
    'delhi': 'Delhi',
    'DELHI': 'Delhi'
})

df['Work_Mode'] = df['Work_Mode'].replace({
    'REMOTE': 'Remote',
    'remote': 'Remote'
})

In [6]:
## Correct Data Types
numeric_cols = ['Age', 'Annual_Salary', 'Experience_Years', 'Performance_Score']

for col in numeric_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce')

df['Joining_Date'] = pd.to_datetime(df['Joining_Date'], errors='coerce')

print(df.dtypes)

Employee_ID                  object
Employee_Name                object
Department           string[python]
Job_Title                    object
Age                         float64
Gender               string[python]
Annual_Salary               float64
Experience_Years            float64
Joining_Date         datetime64[ns]
City                 string[python]
Performance_Score           float64
Work_Mode            string[python]
dtype: object


In [7]:
## Handle Missing Values
numeric_impute_cols = ['Age', 'Annual_Salary', 'Experience_Years', 'Performance_Score']

for col in numeric_impute_cols:
    df[col] = df[col].fillna(df[col].median())

categorical_impute_cols = ['Department', 'Gender', 'City', 'Work_Mode']

for col in categorical_impute_cols:
    df[col] = df[col].fillna(df[col].mode()[0])

print(df.isnull().sum())

Employee_ID          0
Employee_Name        0
Department           0
Job_Title            0
Age                  0
Gender               0
Annual_Salary        0
Experience_Years     0
Joining_Date         0
City                 0
Performance_Score    0
Work_Mode            0
dtype: int64


In [8]:
df['Joining_Date'] = df['Joining_Date'].dt.strftime('%Y-%m-%d')
df['Age'] = df['Age'].round().astype(int)
df['Annual_Salary'] = df['Annual_Salary'].round(2)
df['Experience_Years'] = df['Experience_Years'].round(1)
df['Performance_Score'] = df['Performance_Score'].round().astype(int)

In [9]:
## Verify After Cleaning
print('Final Shape:', df.shape)
print('\nMissing Values:')
print(df.isnull().sum())
print('\nTotal Missing Values:', df.isnull().sum().sum())
print('Duplicate Records:', df.duplicated().sum())
print('\nFinal Data Types:')
print(df.dtypes)
df.head()

Final Shape: (150, 12)

Missing Values:
Employee_ID          0
Employee_Name        0
Department           0
Job_Title            0
Age                  0
Gender               0
Annual_Salary        0
Experience_Years     0
Joining_Date         0
City                 0
Performance_Score    0
Work_Mode            0
dtype: int64

Total Missing Values: 0
Duplicate Records: 0

Final Data Types:
Employee_ID                  object
Employee_Name                object
Department           string[python]
Job_Title                    object
Age                           int64
Gender               string[python]
Annual_Salary               float64
Experience_Years            float64
Joining_Date                 object
City                 string[python]
Performance_Score             int64
Work_Mode            string[python]
dtype: object


,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,Engineering,Data Scientist,48,Other,108371.0,13.2,2017-06-13,Pune,5,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31,Female,108824.0,12.1,2021-09-15,Mumbai,2,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28,Female,119400.0,13.2,2021-02-05,Bengaluru,3,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30,Female,111847.0,9.2,2018-08-05,Jaipur,4,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25,Female,78677.0,10.9,2018-07-24,Hyderabad,4,Hybrid


## Before vs After

| Check | Before | After |
|---|---:|---:|
| Rows | 157 | 150 |
| Missing values | 32 | 0 |
| Duplicate records | 7 | 0 |

In [10]:
df.to_csv('Cleaned_Company_Employee_Dataset.csv', index=False)
print('Cleaned dataset exported successfully.')

Cleaned dataset exported successfully.
